waiter/src/waiter/service_description.clj

```diff
@@ -1252,22 +1253,24 @@
   [reference-type->stale-fn references]
   (every? (fn [reference] (stale-reference? reference-type->stale-fn reference)) references))
 
-(defn service-id->idle-timeout
-  "Computes the idle timeout, in minutes, for a given service.
-   If the service is active or was created by on-the-fly, the idle timeout is retrieved from the service description.
-   Else, the idle timeout is the sum of the fallback period seconds and the stale service timeout."
+(defn service->gc-time
+  "Computes the time when a given service should be GC-ed.
+   If the service has GC disabled (by configuring idle-timeout-mins=0), nil is returned.
+   If the service is active or was created by on-the-fly, the GC time is retrieved from the service description.
+   Else, the GC time is the sum of the fallback period seconds and the stale service timeout."
   [service-id->service-description-fn service-id->references-fn token->token-parameters reference-type->stale-fn
-   attach-token-defaults-fn service-id]
+   attach-token-defaults-fn service-id last-modified-time]
   (let [{:strs [idle-timeout-mins]} (service-id->service-description-fn service-id)
         references (service-id->references-fn service-id)]
     (if (service-references-stale? reference-type->stale-fn references)
       (let [{:strs [stale-timeout-mins]} (attach-token-defaults-fn {})]
-        (log/info service-id "that uses tokens is stale")
+        (log/info service-id "that uses references is stale")
         (if-let [source-tokens (->> references (map :token) (remove nil?) (map :sources) seq)]
-          (source-tokens->idle-timeout token->token-parameters attach-token-defaults-fn source-tokens)
+          (source-tokens->gc-time token->token-parameters attach-token-defaults-fn source-tokens)
           ;; use the default token stale timeout for a stale service built without tokens
-          stale-timeout-mins))
-      idle-timeout-mins)))
+          (t/plus last-modified-time (t/minutes stale-timeout-mins))))
+      (when (pos? idle-timeout-mins)
+        (t/plus last-modified-time (t/minutes idle-timeout-mins))))))
```

Comment:
Shouldn't we still be using `last-request-time` here?

Maybe it makes more sense to have separate functions for computing the normal idle gc time and the eager stale gc time. Then we can take the min of the two.

Or does it make sense for us to extend an instance's life past its idle timeout for token fallback logic? I think I prefer that a token can idle-out as scheduled (since it's getting no traffic!) even if fallback is active. You can imagine a scenario where a service was about to idle-out, but then the token is modified, and the fallback period extends the service's life by several hours. However, I could be convinced that it's more user-friendly to keep the old one around during the fallback period.

Reply:
> Or does it make sense for us to extend an instance's life past its idle timeout for token fallback logic?
> a scenario where a service was about to idle-out, but then the token is modified, and the fallback period extends the service's life by several hours.

This is exactly the behavior we want that triggered this bug fix. A user updating a token expects fallback to help out while the new instance starts up.

> Shouldn't we still be using last-request-time here?

`last-modified-time` is proxy for `last-request-time`, it may be off by a few seconds at worst.

Reply:
> This is the requested behavior that triggered this bug fix. A user updating a token expects fallback to help out while the new instance starts up.

OK, true. It's probably better to extend a service that would have normally idled-out so that we can avoid users being confused about why the old instance shut down too quickly. In practice this is unlikely to happen (a service that has no requests being updated right as it's about to idle-out), and even if it does I think you're right that the possible user confusion is not with the minor resource savings that we'd get by shutting the old service off a tiny bit earlier.